src/random_ray/flat_source_domain.cpp

```diff
@@ -683,4 +803,130 @@ void FlatSourceDomain::output_to_vtk() const
   }
 }
 
+void FlatSourceDomain::apply_external_source_to_source_region(
+  Discrete* discrete, double strength_factor, int64_t source_region)
+{
+  const auto& discrete_energies = discrete->x();
+  const auto& discrete_probs = discrete->prob();
+
+  for (int e = 0; e < discrete_energies.size(); e++) {
+    int g = data::mg.get_group_index(discrete_energies[e]);
+    external_source_[source_region * negroups_ + g] +=
+      discrete_probs[e] * strength_factor;
+  }
+}
+
+void FlatSourceDomain::apply_external_source_to_cell_instances(int32_t i_cell,
+  Discrete* discrete, double strength_factor, int target_material_id,
+  const vector<int32_t>& instances)
+{
+  Cell& cell = *model::cells[i_cell];
+
+  if (cell.type_ != Fill::MATERIAL)
+    return;
+
+  for (int j : instances) {
+    int cell_material_idx = cell.material(j);
+    int cell_material_id = model::materials[cell_material_idx]->id();
+    if (target_material_id == C_NONE ||
+        cell_material_id == target_material_id) {
+      int64_t source_region = source_region_offsets_[i_cell] + j;
+      apply_external_source_to_source_region(
+        discrete, strength_factor, source_region);
+    }
+  }
+}
+
+void FlatSourceDomain::apply_external_source_to_cell_and_children(
+  int32_t i_cell, Discrete* discrete, double strength_factor,
+  int32_t target_material_id)
+{
+  Cell& cell = *model::cells[i_cell];
+
+  if (cell.type_ == Fill::MATERIAL) {
+    vector<int> instances(cell.n_instances_);
+    std::iota(instances.begin(), instances.end(), 0);
+    apply_external_source_to_cell_instances(
+      i_cell, discrete, strength_factor, target_material_id, instances);
+  } else if (target_material_id == C_NONE) {
+    std::unordered_map<int32_t, vector<int32_t>> cell_instance_list =
+      cell.get_contained_cells(0, nullptr);
+    for (const auto& pair : cell_instance_list) {
+      int32_t i_child_cell = pair.first;
+      apply_external_source_to_cell_instances(i_child_cell, discrete,
+        strength_factor, target_material_id, pair.second);
+    }
+  }
+}
+
+void FlatSourceDomain::count_external_source_regions()
+{
+#pragma omp parallel for reduction(+ : n_external_source_regions_)
+  for (int sr = 0; sr < n_source_regions_; sr++) {
+    float total = 0.f;
+    for (int e = 0; e < negroups_; e++) {
+      int64_t se = sr * negroups_ + e;
+      total += external_source_[se];
+    }
+    if (total != 0.f) {
+      n_external_source_regions_++;
+    }
+  }
+}
+
+void FlatSourceDomain::convert_external_sources()
+{
+  // Loop over external sources
+  for (int es = 0; es < model::external_sources.size(); es++) {
+    Source* s = model::external_sources[es].get();
+    IndependentSource* is = dynamic_cast<IndependentSource*>(s);
+    Discrete* energy = dynamic_cast<Discrete*>(is->energy());
```

Comment:
Again, we really should be either asserting the pointer is not null, or using dynamic casts to const references, which will throw an exception if the cast isn't possible.

Reply:
Also, I'm not seeing what necessitates the downcast anyway. From what I see here the only functionality required is in the `Source` base class.

Reply:
Ah, of course! Thanks.